Feature/select

isacoder opened the pull request:
### Requirements

This PR resolves issue #113 that requests a select with two sizes with a label

### Implementation

Sorry for the delay, I misunderstood the requirements since the word "input" was in the description I though this was going to be a full custom component. Then after a conversation with @atomworks made me realize this required a "normal select" with style and not having style for the dropdown options 😓 
 
I hope I have implemented as expected but let me know if this needs to be different.

The hardest part was the placeholder as this is not "native" of the select, I created an invalid option that is shown by default but can't be selected.

I applied the same logic as the Tag with link for this SelectField, when no Label props are sent the label will not be added.

Updated the folders of Storybook because Form was growing big.

### Screenshoots

**Chrome**

<img width="566" alt="Screen Shot 2021-07-14 at 18 42 19" src="https://user-images.githubusercontent.com/10409078/125601911-973327ff-7078-4482-b47d-edcad41b1931.png">

<img width="647" alt="Screen Shot 2021-07-14 at 18 42 27" src="https://user-images.githubusercontent.com/10409078/125601743-d54486c4-684f-441a-b6c5-5ea774730ad7.png">


**Firefox**

<img width="557" alt="Screen Shot 2021-07-14 at 18 41 21" src="https://user-images.githubusercontent.com/10409078/125602009-28446926-2cb3-4581-bc18-df0b456464fb.png">

<img width="666" alt="Screen Shot 2021-07-14 at 18 41 29" src="https://user-images.githubusercontent.com/10409078/125602180-046c3817-2e17-4f21-9962-ffdc2ebac289.png">

atomworks commented on src/Form/atoms/SelectField.tsx:
```diff
@@ -0,0 +1,140 @@
+  labelProps?: ILabel
+  styleSize?: ISelectSizes
+  placeholder?: string
+  selectWidth?: string
```
Can we just use `width` for this as it's the Select component?

isacoder replied:
I was trying to make it flexible like the input, I will just leave it with 100% and let it be adjust to the container.

atomworks replied:
I think when we provide a `<select>` wrapped with the label then it's good to leave the select at 100%. If we're importing it directly I think there is use cases where it makes sense to allow an option width as you've described. 

By default it should just scale automatically but I can see times were it may be used in some flex based columns were we'd want to set the sizing. Perhaps in that case a max-width option would be good too. 

We can iterate later so feel free to make a decision you feel is best and we can adjust in the future as we use it and see how it's being used.

atomworks commented on src/Form/atoms/SelectField.tsx:
```diff
@@ -0,0 +1,140 @@
+
+interface OwnProps {
+  labelProps?: ILabel
+  styleSize?: ISelectSizes
```
Simplify `styleSize` to `size` to stay consistent with how we're doing things in the `Button` component.

isacoder replied:
sorry, I had a problem using size directly in the select because it's a property.
I have moved the value to the container and handled there.

isacoder replied:
Sorry updating again.. so size in the select is the amount of rows allowed to be displayed.
I will keep the interface and the name to avoid confusion.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select

atomworks replied:
Ahhhh, I didn't realise that size was a prop on select already. It's not something I've used much so it didn't occur to me.

Thinking ahead a little, the sizes we use for form fields should always ever be one of two sizes. The standard size (most commonly used) and the compact size. It's not like buttons where we have four sizes. So perhaps an option we can later integrate in to how we handle other fields would be to simply have this as a `boolean` value such as `isCompact`. What do you think?

atomworks commented on src/Form/atoms/SelectField.tsx:
```diff
@@ -0,0 +1,140 @@
+      return prev;
+    });
+
+    const { value } = e.target;
```
Better to keep this at the top of the function for readability.

atomworks commented on src/Form/atoms/SelectField.tsx:
```diff
@@ -0,0 +1,140 @@
+  }
+`;
+
+interface ILabel {
+  htmlFor: string
+  labelText: string
+}
```
I don't think this is consistent with the other form elements we have. This seems like a sensible choice although we need to consider consistency first. What are your thoughts here?

isacoder replied:
I wanted to make having a Label as optional in this component but if label is used I wanted to make htmlFor a required to enforce the developer to provide that value.

We could drop the Label functionality if you want and just have the Select component.

atomworks replied:
Perhaps something like this might be a good middle ground. That way the label is option but if you do provide it then both of those props become mandatory.

```
interface ILabel {
  htmlFor: string
  text: string
}

interface IProps {
    label?: ILabel
}
```

isacoder replied:
I have updated as suggested and added sameRow prop optional to make the label in the same row that the select.

atomworks reviewed (changes requested):
We've chatted quickly on Slack about this but just wanted to recap the hands on test points below (and a couple of other small bugs I noticed). Then there's a few code comments too. Please let me know if anything doesn't make sense. 

- [ ] Width prop isn't working correctly, seems like main container needs the value and internal parts should scale to that.
- [ ] The files that have been moved, can we change that folder name from `InputStories` to just `Input` as we're already in a stories folder so we don't need that distinction.
- [ ] Font weight on placeholder and value seems too heavy.
- [ ] Vertical alignment of font seems off (could be the font weight making it look like this though)
- [ ] Left padding is slightly off.
- [ ] Clicking on the down arrow doesn't register click (and can start a text highlight moment)
- [ ] Make the select text un-text-selectable (related to above)

atomworks reviewed (commented):
I've answered the open questions so I'll submit this as a review to set the  status accordingly. Re-request a review any time you need anything else from me on this.

isacoder commented:
Hi I have updated the story to have a fixwidth option by getting providing the selectWrapper to be used, also added an extra example that is a page list inspired by some of the cases I've sen in the projects.

In the previous update I notice the font-width was 500 but updated to 400 since for lato there is not 500.. but I wasn't sure if this fixed the weight issue, please let me know.

<img width="916" alt="Screen Shot 2021-07-30 at 16 39 06" src="https://user-images.githubusercontent.com/10409078/127619670-f435d641-744d-419f-9f9e-58724e5ae12c.png">

atomworks commented on src/Form/atoms/SelectField.tsx:
```diff
@@ -114,31 +127,31 @@ const SelectField: React.FC<ISelect> = ({
-
-    onChange(value);
-  }, [onChange]);
+    console.log(value);
```
Best to avoid `console.log` in production code.

atomworks reviewed (approved):
Looks good. I've added one comment about a stray `console.log` so if you get a chance to remove that before merging, that would be great. No enough to delay any release though.
